tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp

```diff
@@ -11,104 +11,98 @@
+    static constexpr auto pred              = [](const int x, const int y) { return x == y + 1; };
+
+    static constexpr int good_needle[] = {-1, 0};
+    static constexpr int bad_needle[]  = {0, 0};
+
+    template <ranges::forward_range Fwd1, ranges::forward_range Fwd2>
+    static constexpr void test() {
+        using ranges::find_end, ranges::begin, ranges::end, ranges::iterator_t, ranges::next, ranges::subrange;
+
+        {
+            // Validate range overload [found case]
+            Fwd1 haystack{pairs};
+            Fwd2 needle{good_needle};
+            const same_as<subrange<iterator_t<Fwd1>>> auto result = find_end(haystack, needle, pred, get_first);
+            STATIC_ASSERT(
+                CanMemberSize<subrange<iterator_t<Fwd1>>> == sized_sentinel_for<iterator_t<Fwd1>, iterator_t<Fwd1>>);
+            if constexpr (CanMemberSize<subrange<iterator_t<Fwd1>>>) {
+                assert(result.size() == 2);
+            }
+            assert(result.begin() == next(begin(haystack), 2));
+            assert(result.end() == next(begin(haystack), 4));
+        }
 
-    const auto pred = [](const int x, const int y) { return x == y + 1; };
+        {
+            // Validate range overload [not found case]
+            Fwd1 haystack{pairs};
+            Fwd2 needle{bad_needle};
+            STATIC_ASSERT(
+                CanMemberSize<subrange<iterator_t<Fwd1>>> == sized_sentinel_for<iterator_t<Fwd1>, iterator_t<Fwd1>>);
+            const same_as<subrange<iterator_t<Fwd1>>> auto result = find_end(haystack, needle, pred, get_first);
+            assert(result.empty());
+            assert(result.begin() == end(haystack));
+            assert(result.end() == end(haystack));
+        }
 
-    const array good_needle = {-1, 0};
-    {
-        // Validate range overload [found case]
-        const auto result = find_end(pairs, good_needle, pred, get_first);
-        STATIC_ASSERT(same_as<decltype(result), const subrange<iterator_t<decltype(pairs)>>>);
-        assert(result.size() == 2);
-        assert(result.begin() == pairs.begin() + 4);
-        assert(result.end() == pairs.begin() + 6);
-    }
-    {
-        // Validate iterator + sentinel overload [found case]
-        const auto result =
-            find_end(pairs.begin(), pairs.end(), good_needle.begin(), good_needle.end(), pred, get_first);
-        STATIC_ASSERT(same_as<decltype(result), const subrange<iterator_t<decltype(pairs)>>>);
-        assert(result.size() == 2);
-        assert(result.begin() == pairs.begin() + 4);
-        assert(result.end() == pairs.begin() + 6);
-    }
+        {
+            // Validate iterator + sentinel overload [found case]
+            Fwd1 haystack{pairs};
+            Fwd2 needle{good_needle};
+            const same_as<subrange<iterator_t<Fwd1>>> auto result =
+                find_end(begin(haystack), end(haystack), begin(needle), end(needle), pred, get_first);
+            STATIC_ASSERT(

```

Comment:
Nitpick, is there a reason to shout `STATIC_ASSERT`, at least in ranges code we should be able to be a bit less shouty

Reply:
Yes there is: I refuse to manually map from line number to failure condition when I can make MSVC do it for me by capitalizing `static_assert`.